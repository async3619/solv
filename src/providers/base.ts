import fetch from "node-fetch";

import { Challenge, InputType } from "../utils/types";
import { CachedChallenge } from "../utils/retrieveChallenge";

export abstract class BaseProvider {
    public abstract checkUrl(url: string): boolean;
    public abstract retrieve(url: string): Promise<Challenge>;
    public abstract getId(url: string): string | number;

    protected constructor(public readonly needJs: boolean) {}

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public hydrateCache({ version, ...challenge }: CachedChallenge): Challenge {
        return {
            ...challenge,
            provider: this,
        };
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem !== "string" ? JSON.stringify(outputItem) : outputItem;
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public beforeExecute(sourceCode: string, input: InputType): string {
        return sourceCode;
    }

    public getName(): string {
        return this.constructor.name.replace("Provider", "");
    }

    protected async fetch(url: string): Promise<string> {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch (${response.status}) - ${response.statusText}`);
        }

        return response.text();
    }
}
