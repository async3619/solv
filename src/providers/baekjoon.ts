import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";
import { Challenge } from "../types";

export class BaekjoonProvider extends BaseProvider {
    private readonly urlRegex = /^https?:\/\/(www\.)?acmicpc\.net\/problem\/([0-9]*?)$/;

    public checkUrl(url: string) {
        return this.urlRegex.test(url);
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        const htmlCode = await this.fetch(url);
        const document = parseDOM(htmlCode);

        const titleDOM = document.querySelector("#problem_title");
        const descriptionDOM = document.querySelector("#problem_description");
        const inputDescriptionDOM = document.querySelector("#problem_input");
        const outputDescriptionDOM = document.querySelector("#problem_output");
        const ioItem = document.querySelectorAll(".sampledata");

        const ioText: [boolean, string][] = [];
        for (const item of ioItem) {
            ioText.push([item.id.includes("-input-"), item.innerHTML.trim()]);
        }

        if (
            !titleDOM?.textContent ||
            !descriptionDOM?.textContent ||
            !inputDescriptionDOM?.textContent ||
            !outputDescriptionDOM?.textContent
        ) {
            throw new Error("Failed to parse challenge information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            inputDescription: inputDescriptionDOM.textContent.trim(),
            outputDescription: outputDescriptionDOM.textContent.trim(),
            input: ioText.filter(([i]) => i).map(([, t]) => t),
            output: ioText.filter(([i]) => !i).map(([, t]) => t.replace(/\r\n/g, "\n")),
            provider: this,
            initialCode: `
function solution(input: string[]) {
    // ...
}

(() => {
    const readline = require("readline");

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const input: string[] = [];
    rl.on("line", line => {
        input.push(line);
    }).on("close", function () {
        solution(input);
        process.exit();
    });
})();
`.trim(),
        };
    }
}
