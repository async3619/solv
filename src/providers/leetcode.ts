import { assertParse, is } from "typia";
import { parse as parseDOM } from "node-html-parser";
import { Client, CombinedError, createClient } from "@urql/core";
import { stripHtml } from "string-strip-html";

import { BaseProvider } from "./base";

import { Challenge, InputType } from "../utils/types";
import { LeetCodeQuestionDocument } from "../queries/leetcode";
import type { LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables } from "../queries.data";
import { QuestionMetadata } from "./leetcode.types";

const URL_CHECK_REGEX = /(https?:\/\/)?leetcode\.com\/problems\/([a-z0-9-]+)\/?$/;

export class LeetCodeProvider extends BaseProvider {
    private readonly client: Client;

    public constructor() {
        super(false);

        this.client = createClient({
            url: "https://leetcode.com/graphql",
            requestPolicy: "network-only",
            fetchOptions: () => {
                return {
                    method: "GET",
                    headers: {
                        "content-type": "application/json",
                    },
                };
            },
        });
    }

    public checkUrl(url: string): boolean {
        return URL_CHECK_REGEX.test(url);
    }

    public getId(url: string): string | number {
        const match = URL_CHECK_REGEX.exec(url);
        if (!match) {
            throw new Error("Failed to parse challenge ID.");
        }

        return match[2];
    }

    public beforeExecute(sourceCode: string, input: InputType, challenge: Challenge): string {
        let entryPointName = "solution";
        if (challenge.initialCode) {
            const match = challenge.initialCode.match(/function ([a-zA-Z0-9]+)\(/);
            if (match) {
                entryPointName = match[1];
            }
        }

        return [sourceCode, `console.info(${entryPointName}(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const titleSlug = this.getId(url);
        if (!is<string>(titleSlug)) {
            throw new Error("Failed to parse challenge ID.");
        }

        try {
            const { error, data } = await this.client
                .query<LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables>(LeetCodeQuestionDocument, { titleSlug })
                .toPromise();

            if (error) {
                throw error;
            }

            if (!data?.question) {
                throw new Error("Failed to retrieve challenge data.");
            }

            const codeSnippet = data.question.codeSnippets.find(item => item.langSlug === "typescript");
            if (!codeSnippet) {
                throw new Error("This challenge does not support TypeScript.");
            }

            const document = parseDOM(data.question.content);
            const testCases = [...document.querySelectorAll("pre")].filter(item => {
                if (!item.parentNode) {
                    return false;
                }

                if (!item.previousElementSibling) {
                    return false;
                }

                const title = item.previousElementSibling.querySelector("strong.example");
                if (!title) {
                    return false;
                }

                return true;
            });

            const metadata = assertParse<QuestionMetadata>(data.question.metaData);
            const testCaseSummaries = testCases.map(item => stripHtml(item.textContent.trim()).result);
            const inputs: InputType[] = [];
            const outputs: string[] = [];
            for (const summary of testCaseSummaries) {
                const [input, output] = summary.split("\n").map(item => {
                    return item.trim().replace(/^[A-Za-z]*?: /, "");
                });

                const inputParameters = input.split(/, /).map(item => item.split(" = "));
                const data: InputType = [];
                for (const [key, value] of inputParameters) {
                    const index = metadata.params.findIndex(item => item.name === key);
                    if (index === -1) {
                        throw new Error(`Failed to find parameter "${key}".`);
                    }

                    data[index] = value;
                }

                inputs.push(data);
                outputs.push(output.replace("Output: ", ""));
            }

            return {
                id: data.question.questionId,
                title: data.question.questionTitle,
                description: stripHtml(data.question.content).result,
                provider: this,
                initialCode: codeSnippet.code,
                input: inputs,
                output: outputs,
            };
        } catch (e) {
            if (e instanceof CombinedError) {
                if (e.networkError) {
                    throw e.networkError;
                } else if (e.graphQLErrors && e.graphQLErrors.length > 0) {
                    throw e.graphQLErrors[0];
                }
            }

            throw e;
        }
    }
}
