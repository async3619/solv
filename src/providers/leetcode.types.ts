export interface QuestionMetadata {
    name: string;
    params: Array<{
        name: string;
        type: string;
    }>;
    return: {
        type: string;
    };
}
