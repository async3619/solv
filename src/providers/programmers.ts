import { parse as parseDOM } from "node-html-parser";

import { BaseProvider } from "./base";

import { findSiblingElement } from "../utils/findSiblingElement";
import { parseSampleTable } from "../utils/parseSampleTable";
import { Challenge } from "../utils/types";

export class ProgrammersProvider extends BaseProvider {
    public constructor() {
        super(false);
    }

    public checkUrl(url: string): boolean {
        return /(https?:\/\/)?(school\.)?programmers\.co\.kr\/learn\/courses\/([0-9]*?)\/lessons\/[0-9]([0-9]*?)$/.test(
            url,
        );
    }

    public serializeOutput(outputItem: any): string {
        return typeof outputItem === "string" ? `"${outputItem}"` : JSON.stringify(outputItem);
    }

    public beforeExecute(sourceCode: string): string {
        return [sourceCode, `console.info(solution(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const targetUrl = new URL(url);
        targetUrl.searchParams.set("language", "javascript");

        const htmlCode = await this.fetch(targetUrl.toString());

        const document = parseDOM(htmlCode);
        const descriptionDOM = document.querySelector("#tour2 > div > div");
        const titleDOM = document.querySelector("#tab > li");
        if (!titleDOM?.textContent || !descriptionDOM?.textContent) {
            throw new Error("Failed to parse challenge information.");
        }

        const headers = document.querySelectorAll(".guide-section h5");
        const ioSampleHeaderRegex = /^입출력 ?예$/;
        const ioSampleHeaderDOM = headers.find(header => ioSampleHeaderRegex.test(header.textContent.trim()));
        if (!ioSampleHeaderDOM) {
            throw new Error("Failed to parse challenge input / output samples.");
        }

        const ioSampleTable = findSiblingElement(ioSampleHeaderDOM, node => node.tagName === "TABLE");
        if (!ioSampleTable) {
            throw new Error("Failed to find challenge output sample table.");
        }

        const initialCode = document
            .querySelector(`input[type="hidden"][name^="initial_code_"]`)
            ?.getAttribute("value");

        if (!initialCode) {
            throw new Error("Failed to parse initial code information.");
        }

        const id = targetUrl.pathname.split("/").at(-1);
        if (!id) {
            throw new Error("Failed to retrieve challenge id.");
        }

        return {
            ...parseSampleTable(ioSampleTable),
            id,
            title: titleDOM.textContent.trim(),
            description: descriptionDOM.textContent.trim(),
            provider: this,
            initialCode,
        };
    }
}
