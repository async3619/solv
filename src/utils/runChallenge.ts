import * as Listr from "listr";
import * as chalk from "chalk";

import { Challenge, Config, InputOutput, InputType } from "./types";

import { BaseProvider } from "../providers/base";

import { breakLine, renderSection } from "./cli";
import { TestCaseFailedError } from "./TestCaseFailedError";
import { transpileAndRun } from "./transpileAndRun";
import { normalizeString } from "./normalizeString";

export async function runTestCase(
    index: number,
    input: InputType,
    output: string,
    targetPath: string,
    provider: BaseProvider,
    isCustom = false,
) {
    try {
        const data = await transpileAndRun(input, output, targetPath, provider);
        if (data instanceof Error) {
            throw data;
        }

        if (data[0] !== output) {
            throw new TestCaseFailedError(index, data, undefined, isCustom);
        }
    } catch (e) {
        if (e instanceof TestCaseFailedError) {
            throw e;
        } else if (e instanceof Error) {
            throw new TestCaseFailedError(
                index,
                [(e as any).stderr || e.stack, ""],
                "Test case failed with an error",
                isCustom,
            );
        }
    }
}

export async function runChallenge(challenge: Challenge, targetPath: string, config: Config | null) {
    breakLine();

    const items: InputOutput[] = [];
    if (config) {
        const cases = config.cases
            .filter(
                ({ id, provider }) =>
                    id.toString() === challenge.id.toString() &&
                    challenge.provider.getName().toLowerCase() === provider,
            )
            .map(p => p.items)
            .flat();

        items.push(
            ...cases.map<InputOutput>(({ input, output }) => ({
                input: normalizeString(input),
                output: normalizeString(output),
                isCustom: true,
            })),
        );
    }

    items.push(
        ...challenge.input.map<InputOutput>((input, i) => ({
            input,
            output: challenge.output[i],
        })),
    );

    let currentTestCaseIndex = 0;
    const instance = new Listr([
        ...items.map(({ input, output, isCustom }, index) => ({
            title: `Running with ${isCustom ? "custom " : ""}test case #${index + 1}`,
            task: () => runTestCase(index, input, output, targetPath, challenge.provider, isCustom),
        })),
    ]);

    try {
        await instance.run();
    } catch (e) {
        let result = [""];
        if (e instanceof TestCaseFailedError) {
            currentTestCaseIndex = e.testCaseIndex;
            result = e.result;
        }

        const { input, output } = items[currentTestCaseIndex];

        breakLine(2);
        renderSection("Input", input, chalk.cyan);
        renderSection("Output (expected)", output, chalk.green);
        renderSection("Output", result[0], chalk.red);

        if (result.length > 1 && result[1]) {
            renderSection("Debug", result[1], chalk.magenta);
        }
    }
}
