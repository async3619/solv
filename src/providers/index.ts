import { BaseProvider } from "./base";

import { BaekjoonProvider } from "./baekjoon";
import { ProgrammersProvider } from "./programmers";
import { LeetCodeProvider } from "./leetcode";

const TARGET_PROVIDERS: BaseProvider[] = [new BaekjoonProvider(), new ProgrammersProvider(), new LeetCodeProvider()];

export function getProvider(targetUrl: string) {
    const provider = TARGET_PROVIDERS.find(provider => provider.checkUrl(targetUrl));
    if (!provider) {
        throw new Error("there was no matched coding challenge service provider on records. abort.");
    }

    return provider;
}
