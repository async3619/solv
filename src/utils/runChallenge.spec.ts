import * as fs from "fs-extra";

import { BaekjoonProvider } from "../providers/baekjoon";

import { runChallenge, runTestCase } from "./runChallenge";

const SAMPLE_CODE = `
function solution(input: string[]) {
    console.log(input);
}

(callback => {
    if (typeof process !== "undefined" && "env" in process && "arguments" in process.env && process.env.arguments) {
        solution(process.env.arguments);
        return;
    }

    const readline = require("readline");
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const input: string[] = [];
    rl.on("line", line => {
        input.push(line);
    }).on("close", function () {
        callback(input);
        process.exit();
    });
})(solution);
`.trim();

describe("runTestCase", () => {
    it("should runs given source code file", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(SAMPLE_CODE);
        (fs.writeFile as any) = () => Promise.resolve();

        await runTestCase(0, "Test", "Test", "./test.ts", new BaekjoonProvider());
    });

    it("should throws an error when output is not match", async () => {
        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(SAMPLE_CODE);
        (fs.writeFile as any) = () => Promise.resolve();
        expect.assertions(1);

        await expect(runTestCase(0, "Test", "Wrong Answer", "./test.ts", new BaekjoonProvider())).rejects.toThrow(
            "Output was not matched with test case #1",
        );
    });

    it("should throws an error when given source thrown", async () => {
        jest.mock("fs-extra");

        (fs.readFile as any) = () =>
            Promise.resolve(`
                throw new Error("Test Error.");
            `);

        (fs.writeFile as any) = () => Promise.resolve();
        expect.assertions(1);

        await expect(runTestCase(0, "Test", "Wrong Answer", "./test.ts", new BaekjoonProvider())).rejects.toThrow(
            "Test case failed with an error",
        );
    });
});

describe("runChallenge", () => {
    const installHook = (code: string) => {
        const buffer: string[] = [];
        console.log = (content: string) => buffer.push(content);
        console.info = (content: string) => buffer.push(content);

        jest.mock("fs-extra");
        (fs.readFile as any) = () => Promise.resolve(code);
        (fs.writeFile as any) = () => Promise.resolve();

        return buffer;
    };

    it("should runs given challenge properly", async () => {
        const buffer: string[] = installHook(SAMPLE_CODE);
        await runChallenge(
            {
                id: 0,
                title: "Test",
                provider: new BaekjoonProvider(),
                input: ["Test"],
                output: ["Output"],
            },
            "./test.ts",
            null,
        );

        expect(buffer.join("\n").replace(/\[[0-9]{2}:[0-9]{2}:[0-9]{2}\]/g, "")).toMatchSnapshot();
    });

    it("should shows error message whenever test case throws an error", async () => {
        const buffer: string[] = installHook(`throw new Error("Test Error.");`);
        await runChallenge(
            {
                id: 0,
                title: "Test",
                provider: new BaekjoonProvider(),
                input: ["Test"],
                output: ["Output"],
            },
            "./test.ts",
            null,
        );

        expect(buffer.join("\n").replace(/\[[0-9]{2}:[0-9]{2}:[0-9]{2}\]/g, "")).toMatchSnapshot();
    });

    it("should shows debug message whenever test case leave debugging log", async () => {
        const buffer: string[] = installHook(`console.debug("Test Log")`);
        await runChallenge(
            {
                id: 0,
                title: "Test",
                provider: new BaekjoonProvider(),
                input: ["Test"],
                output: ["Output"],
            },
            "./test.ts",
            null,
        );

        expect(buffer.join("\n").replace(/\[[0-9]{2}:[0-9]{2}:[0-9]{2}\]/g, "")).toMatchSnapshot();
    });

    it("should runs custom test case only when its given", async () => {
        const buffer: string[] = installHook(SAMPLE_CODE);
        await runChallenge(
            {
                id: 0,
                title: "Test",
                provider: new BaekjoonProvider(),
                input: ["Test"],
                output: ["Output"],
            },
            "./test.ts",
            {
                cases: [
                    {
                        id: 0,
                        provider: "baekjoon",
                        items: [
                            {
                                input: "Test",
                                output: "Output",
                            },
                        ],
                    },
                ],
            },
        );

        expect(buffer.join("\n").replace(/\[[0-9]{2}:[0-9]{2}:[0-9]{2}\]/g, "")).toMatchSnapshot();
    });
});
