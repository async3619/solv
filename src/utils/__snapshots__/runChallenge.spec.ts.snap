// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runChallenge should runs custom test case only when its given 1`] = `
"[2m[22m Running with custom test case #1 [started]
[2m[22m Running with custom test case #1 [failed]
[2m[22m â†’ Output was not matched with custom test case #1


Input
===============
[36mTest[39m

Output (expected)
===============
[32mOutput[39m

Output
===============
[31mTest[39m"
`;

exports[`runChallenge should runs given challenge properly 1`] = `
"[2m[22m Running with test case #1 [started]
[2m[22m Running with test case #1 [failed]
[2m[22m â†’ Output was not matched with test case #1


Input
===============
[36mTest[39m

Output (expected)
===============
[32mOutput[39m

Output
===============
[31mTest[39m"
`;

exports[`runChallenge should shows debug message whenever test case leave debugging log 1`] = `
"[2m[22m Running with test case #1 [started]
[2m[22m Running with test case #1 [failed]
[2m[22m â†’ Output was not matched with test case #1


Input
===============
[36mTest[39m

Output (expected)
===============
[32mOutput[39m

Output
===============
[31m[3m(empty)[23m[39m

Debug
===============
[35mTest Log[39m"
`;

exports[`runChallenge should shows error message whenever test case throws an error 1`] = `
"[2m[22m Running with test case #1 [started]
[2m[22m Running with test case #1 [failed]
[2m[22m â†’ Test case failed with an error


Input
===============
[36mTest[39m

Output (expected)
===============
[32mOutput[39m

Output
===============
[31mError: Test Error.[39m"
`;
